plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/organize/imports/ImportsComputer.xtend

```diff
@@ -145,271 +148,213 @@ public class ImportsComputer {
 
 		// Ask user to disambiguate things:
-		val Map<String, Collection<ImportableObject>> ambiguousSolution = resolution.asMap.filter[p1, p2|p2.size > 1];
+		val Map<String, Collection<ImportableObject>> ambiguousSolution = resolutions.asMap.filter[p1, p2|p2.size > 1];
+
 		val Multimap<String, ImportableObject> forDisambiguation = LinkedHashMultimap.create();
 		ambiguousSolution.forEach[p1, p2|forDisambiguation.putAll(p1, p2)];
 
-		// add potential solutions for still broken names
-		namesThatWeBroke.forEach [ brokenName |
-			// if there is disambiguation pending for a given name, don't duplicate solutions
-			if (!forDisambiguation.keySet.contains(brokenName)) {
-				val idSolutions = newLinkedHashSet(); // newArrayList();
-				// TODO how to decide on which we use?
-				/*
-				 * Should we add both, or pick one? Or add other if the first one did not find anything?
-				 */
-				idSolutions.addAll(scopeIdRef.mapToImportableObjects(brokenName));
-				idSolutions.addAll(scopeTypeRef.mapToImportableObjects(brokenName));
-				forDisambiguation.putAll(brokenName, idSolutions);
+		// try to automatically disambiguate
+		val resolved = new HashSet<String>();
+		forDisambiguation.keySet.forEach [ key |
+			val multiSolutions = forDisambiguation.get(key);
+			val containsOnlyNamespaces = multiSolutions.filter[!it.asNamespace].empty
+			if (containsOnlyNamespaces) {
+				ret.add(importsFactory.createImport(multiSolutions.head, contextProject, nodelessMarker))
+				resolved.add(key)
 			}
-		];
+		]
+		resolved.forEach [ key |
+			forDisambiguation.removeAll(key)
+		]
 
-		val chosenSolutions = disambiguate(forDisambiguation, interaction);
+		//automatic / semi-automatic / or user only disambiguation
+		val chosenSolutions = DisambiguateUtil.disambiguate(forDisambiguation, interaction,
+			importProvidedElementLabelprovider);
 
-		chosenSolutions.forEach[ret.add(importsFactory.createImport(it, nodelessMarker))];
+		chosenSolutions.forEach[ret.add(importsFactory.createImport(it, contextProject, nodelessMarker))];
 
 		return ret;
 	}
 
-	/** Filters scope by given name and maps result to importable objects. */
-	private def Iterable<ImportableObject> mapToImportableObjects(IScope scopeIdRef, String brokenName) {
-		scopeIdRef
-			.allElements
-			.filter[candidateFilter.apply(it)]
-			.filter[it.name.lastSegment == brokenName]
-			.map[new ImportableObject(brokenName, it, false)]
+	/** Finds unresolved cross references in this script and combines them with provided broken names. 
+	 * @returns list of resolutions for all broken names.
+	 */
+	private def Multimap<String, ImportableObject> createResolutionsForBrokenNames(Script script,
+		IN4JSProject contextProject, Set<String> namesThatWeBroke) {
+		val Multimap<String, ImportableObject> resolutions = LinkedHashMultimap.create();
+
+		val Iterable<ReferenceProxyInfo> unresolved = script.findProxyCrossRefInfo.filter[referenceFilter.test(it)].
+			filter[it.eobject instanceof MemberAccess === false]
```

Comment:
Why are you filtering out MemberAccess? MemberAccess was introduced shortly to denote AST nodes that has a cached composed member. For instance `ParameterizedPropertyAccess` is a `MemberAccess`. Why is `MemberAccess` related to organize imports?

Reply:
To reduce lookups we do, especially that we cannot do anything about property that is accessed, we can only resolve identifier on which it is resolved. For example
```typescript
// file1
export public class X {
	public f = "x.f"
}
export public var y : X;
```

```typescript
//file2
y.f; // Couldn't resolve reference to IdentifiableElement 'y'.
y.x; // Couldn't resolve reference to IdentifiableElement 'y'.
```
we get all unresolved references, which in this case will be two `IdentifierRef` (for `y`) and we will get two `ParametrizedPropertyAccessExpression` (for `.f` and `.x`). 
![1](https://user-images.githubusercontent.com/944287/30214249-fc5a549e-94ab-11e7-8641-50c34398ded9.png)
We ignore all `MemberAccess`, (in this case those`ParametrizedPropertyAccessExpression`) and we are just interested in their targets (i.e. for `IdentifierRef`). 

After organising we get
```typescript
//file2
import {y} from "file1";

y.f; //y is resolved, y.f is resolved
y.x; //y is resolved, y.x is not resolved
```
![2](https://user-images.githubusercontent.com/944287/30214308-35048486-94ac-11e7-9cae-ba5a45a9ad14.png)

Reply:
Be careful: not only ParameterizedAccess is a MemberAccess but IndexedExpression, JSXPropertyAttribute are also MemberAccess. Please double check!